### PR TITLE
Samtools target read count sampling

### PR DESCRIPTION
**Pull request by bwlang**
FOR CONTRIBUTOR:
* [x] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [x] - License permits unrestricted use (educational + commercial)
* [ ] - This PR adds a new tool or tool collection
* [x] - This PR updates an existing tool or tool collection
* [ ] - This PR does something else (explain below)

FOR REVIEWER:
* [x] .shed.yml file ok
    - [x] Toolshed user `iuc` has access to associated toolshed repo(s)
* [x] Indentation is correct (4 spaces)
* [x] Tool version/build ok
* [x] `<command/>`
  - [x] Text parameters, input and output files `'single quoted'`
  - [x] Use of `<![CDATA[ ... ]]>` tags
  - [x] Parameters of type `text` or having `optional="true"` attribute are checked with `if str($param)` before being used
* [x] Data parameters have a `format` attribute containing datatypes recognised by Galaxy
* [x] Tests
  - [x] Parameters are reasonably covered
  - [x] Test files are appropriate
* [x] Help
  - [x] Valid restructuredText and uses `<![CDATA[ ... ]]>` tags
* [x] Complies with other best practice in [Best Practices Doc](http://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html)

### CODE REVIEW

**Comment by bwlang** [comment]
hmm

samtools view -@ 0 -s 7.2 -h -o outfile --output-fmt-option no_ref infile

output on linux: 
```
@HD	VN:1.4	SO:unsorted
@SQ	SN:CHROMOSOME_I	LN:100
@RG	ID:UNKNOWN	SM:UNKNOWN
@PG	ID:bowtie2	PN:bowtie2	VN:2.0.0-beta5
@PG	ID:0	CL:aaaaa/aaa/aaaaa/aaaaaa/aaaaaaaaa/aaa/iuc/package_aaaaaaaaa_x_y/aaaaaaaaaaaa/bin/aaaaaaaaaaaaaaaaa aaaaaaaaaa /aaaa/aaaaa/aaa/aaaaaaaaaaaaaaaaaaa/tools/aaaaaaaaa/test-data/test.cram aa /aaaa/aaaaa/aaa/aaaaaaaaaaaaaaaaaaa/tools/aaaaaaaaa/test-data/test.fa -O test	PN:samtools	VN:1.2
SRR065390.14978392	16	CHROMOSOME_I	2	1	27M1D73M	*	0	0	CCTAGCCCTAACCCTAACCCTAACCCTAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAA	#############################@B?8B?BA@@DDBCDDCBC@CDCDCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC	RG:Z:UNKNOWN	XG:i:1	XM:i:5	XN:i:0	XO:i:1	AS:i:-18	XS:i:-18	YT:Z:UU
SRR065390.921023	16	CHROMOSOME_I	3	12	100M	*	0	0	CTAAGCCTAAATCTAAGCCTAACCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAA	###############################################???88:;98768700000<>:BBA?BBAB?BBBBBBBB>B>BB::;?:00000	RG:Z:UNKNOWN	XG:i:0	XM:i:3	XN:i:0	XO:i:0	AS:i:-6	XS:i:-13	YT:Z:UU
SRR065390.6023338	0	CHROMOSOME_I	3	1	100M	*	0	0	CTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAAGCTAC	CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC@CCDDDBCCABB=DABBA?################	RG:Z:UNKNOWN	XG:i:0	XM:i:3	XN:i:0	XO:i:0	AS:i:-6	XS:i:-6	YT:Z:UU
```
sam command on mac:
```
@HD	VN:1.4	SO:unsorted
@SQ	SN:CHROMOSOME_I	LN:100
@RG	ID:UNKNOWN	SM:UNKNOWN
@PG	ID:bowtie2	PN:bowtie2	VN:2.0.0-beta5
@PG	ID:0	CL:aaaaa/aaa/aaaaa/aaaaaa/aaaaaaaaa/aaa/iuc/package_aaaaaaaaa_x_y/aaaaaaaaaaaa/bin/aaaaaaaaaaaaaaaaa aaaaaaaaaa /aaaa/aaaaa/aaa/aaaaaaaaaaaaaaaaaaa/tools/aaaaaaaaa/test-data/test.cram aa /aaaa/aaaaa/aaa/aaaaaaaaaaaaaaaaaaa/tools/aaaaaaaaa/test-data/test.fa -O test	PN:samtools	VN:1.2
SRR065390.1871511	16	CHROMOSOME_I	3	1	100M	*	0	0	CTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAA	<?@<@A8>0:BB@>B<=B@???@=8@B>BB@CA@DACDCBBCCCA@CCCCACCBCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC	RG:Z:UNKNOWN	XG:i:0	XM:i:0	XN:i:0	XO:i:0	AS:i:0	XS:i:0	YT:Z:UU
SRR065390.3743423	16	CHROMOSOME_I	3	1	100M	*	0	0	CTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAAGCCTAA	##################?6@:7<=@3=@ABAAB>BDBBABADABDDDBDDBCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCCC	RG:Z:UNKNOWN	XG:i:0	XM:i:0	XN:i:0	XO:i:0	AS:i:0	XS:i:0	YT:Z:UU
```

I think this cross platorm inconsistency might be a bug in samtools... anybody seen something like that before? Probably not important enough to fix... i'll just change the expected output to the unix version so the tests will pass reliably

**Comment by peterjc** [comment]
This is rather clever - I like it, hiding some command line magic behind a Galaxy wrapper.

I wrote a Python script for doing (reproducible) read sampling instead (which supports other file formats), and had been meaning to add SAM/BAM support to it:

https://github.com/peterjc/pico_galaxy/tree/master/tools/sample_seqs
http://toolshed.g2.bx.psu.edu/view/peterjc/sample_seqs

**Comment by bwlang** [comment]
I generally prefer obvious over clever - but I'll take what I can get ;)

File: tool_collections/samtools/samtools_view/samtools_view.xml [hunk 1]
```diff
@@ -11,7 +11,19 @@
...
-
+        #if $cond_subsample.select_subsample == 'target':
+            ##this must be done before the main command because we don't know the total # reads until execution time
+            #if str($cond_subsample.seed) == '':
+                #set seed="int(rand() * 32768)"
+            #else
+                #set seed=$cond_subsample.seed
+            #end if
+            #if $input.is_of_type('sam'):
+                sample_fragment=`samtools view -c infile | awk '{s=\$1} END {frac=$cond_subsample.target/s; print(frac < 1 ? "-s "  $seed+frac : "")}'` &&
```
Comment (mvdbeek) [inline tool_collections/samtools/samtools_view/samtools_view.xml]: We should add gawk to the requirements then

Comment (bwlang) [reply tool_collections/samtools/samtools_view/samtools_view.xml]: @mvdbeek  good catch... and I got to learn about yield too :)

File: tool_collections/samtools/samtools_view/samtools_view.xml [hunk 2]
```diff
@@ -3,15 +3,29 @@
...
     <command><![CDATA[
         @ADDTHREADS@
         ## prepare reference data
         @PREPARE_FASTA_IDX@
         @PREPARE_IDX@
-
+        #if $cond_subsample.select_subsample == 'target':
+            ##this must be done before the main command because we don't know the total # reads until execution time
+            #if str($cond_subsample.seed) == '':
+                #set seed="int(rand() * 32768)"
```
Comment (bernt-matthias) [inline tool_collections/samtools/samtools_view/samtools_view.xml]: dont you need an import for using rand()? 

otherwise use bash $RANDOM

**Comment by bernt-matthias** [comment]
Wondering if you get the exact number of reads? 

As far as I remember samtools decides for each read(pair) with a certain probability if it is kept / not, where the probability is equal to the selected fraction. 

If I'm right, neither an exact fraction nor an exact number can be guaranteed...

**Comment by bwlang** [comment]
@bernt-matthias 
Depending on the bam file size you get approximately the targeted number of reads, but not an exact number (hopefully "target" conveys that). I think that's fine for most purposes.

I tried using \$RANDOM first, but I struggled to get awk to call $RANDOM from the shell and I wasn't sure about bash.  Awk has rand by default.

**Comment by bernt-matthias** [comment]
Yep. That's what I thought. And its perfectly fine. 

Something along the lines of `awk -v randvalue=\$RANDOM '{...}'` should work

**Comment by bwlang** [comment]
I guess I'm not sure what switching to $RANDOM gets us... 
Still need the conditional, still need awk.
Maybe it's not clear this way that rand() is an awk function?
I just pushed a change to address that.

Do you think I should change the implementation to $RANDOM instead? I think that looks a bit less readable, but I don't have a strong opinion about it.

**Comment by bernt-matthias** [comment]
Ahh, then its fine.

**Comment by bwlang** [comment]
great - please let me know if other changes are needed before merging.

**Comment by bernt-matthias** [comment]
nice :+1:

**Comment by bernt-matthias** [comment]
Even if unrelated to this PR, there is a bug (sorry):

https://github.com/galaxyproject/tools-iuc/blob/8d879e87af15d9fc4ae82850e79d8fe120e6f176/tool_collections/samtools/samtools_view/samtools_view.xml#L113

quotes don't work here. 

See https://github.com/galaxyproject/tools-iuc/pull/1972#issuecomment-424684378 and below. Would be really nice if you could adapt the new region macros and tokens that are introduced by this PR.

**Comment by wm75** [comment]
@bwlang Do you remember why you chose to do this via a call of idxstats for everything but SAM instead of using view -c unconditionally?

**Comment by wm75** [comment]
@bwlang Sorry, it's obviously much faster to just reuse the index if it exists. It's late and I should really stop working I guess.

**Pull request merged**